sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -142,14 +140,21 @@ object ScanOperation extends OperationHelper with PredicateHelper {
       case Filter(condition, child) =>
         collectProjectsAndFilters(child) match {
           case Some((fields, filters, other, aliases)) =>
-            // Follow CombineFilters and only keep going if the collected Filters
-            // are all deterministic and this filter doesn't have common non-deterministic
+            // Follow CombineFilters and only keep going if 1) the collected Filters
+            // and this filter are all deterministic or 2) if this filter is non-deterministic,
+            // but it's the only one who doesn't have common non-deterministic
             // expressions with lower Project.
-            if (filters.forall(_.deterministic) &&
-              !hasCommonNonDeterministic(Seq(condition), aliases)) {
+            if (filters.nonEmpty && filters.forall(_.deterministic)) {
```

Comment:
I think this should be `filters.nonEmpty && filters.forall(_.deterministic) && condition.deterministic && ! hasCommonNonDeterministic...`

Reply:
`condition` is checked below.

Reply:
how about
```
val canMergeFilters = (filters.nonEmpty && filters.forall(_.deterministic) && condition.deterministic) || filters.isEmpty
if (canMergeFilters && !hasCommonNonDeterministic(...)) {
  ...
}
```

Reply:
looks good!